### ISSUE TITLE
Had to Force TLS 1.2 in MakeAPIRequest

### ISSUE THREAD

**Issue by atguser** [bcemmett/SurveyMonkeyApi-v3#66]
Our downloads failed last night with a "The request was aborted: Could not create SSL/TLS secure channel." error.  We got *our* library back up by updating MakeApiRequest() in the SurveyMonkeyAPI.Infrastructure class:
    ServicePointManager.Expect100Continue = true;
    ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;

This isn't as robust as I'd like, but it works.  Are there any better solutions out there?

**Comment by Dan-s-nmd** [comment]
Thanks, I had to use your solution as well.  I was using a nuget dll before, so I had to eliminate that and pull the current code and include it in my project.  This became quite a hassle as I was working on code from MSVS 2012 Several of the code functions did not translate in 2012, had to fix the null checks (?.) and the Arrow functions => before I could use the code.  It would be really helpful if bcemmett could update the API as it will likely no longer be used from nuget if this fix is not made.

**Comment by bcemmett** [comment]
Thanks for reporting this. It would have been great if they'd let developers know in advance of just making that change! I think this only affects older versions of Windows (not sure what the threshold is) but it also broke something I have on a Win 2008 Server machine.

My understanding is that `ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;` is a setting which gets applied globally across the entire appdomain. So you shouldn't need to set this from within the SurveyMonkeyApi library - if you just include those two lines anywhere in the code which uses the library it should have the same effect. @Dan-s-nmd did you happen to try that?

It's tricky to know what to do here - if the library changes those settings, it would also impact ANY other web requests the parent application makes. So for example if that app is also dealing with some web service that requires older security settings, making the change within this library would break usage of those other apps. Maybe that's an edge case that's worth accepting - am not sure. Need to read a bit about this and then sleep on it I think.

The only way I can see to keep this isolated is something like having this library create a new appdomain and run the web client in that using .net remoting, but that feels quite drastic.

Any thoughts would be very welcome!

**Comment by atguser** [comment]
I found a more advanced hack on StackOveflow that allows you to specify a list of protocol types using pipes: `System.Net.ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls | SecurityProtocolType.Tls11 | SecurityProtocolType.Tls12;`

**Comment by Dan-s-nmd** [comment]
@bcemmett,  Just tried it a few minutes ago.  You are correct the servicepointmanager is rather global.  

If you decide not to make the update, I would make sure that something is added in both the readme and the Nuget dll repository. Otherwise they are going to think it is broken.   

Maybe add a function that is called that stores the current security protocol, when the API connection request is called, and then changes it to tls 1.2 then once the call is over reverts to the stored value?   SecurityprotocolType is enumerable so that should prove relatively easy. 

@atguser  I used the same trick from SO it is in my first comment.

**Comment by atguser** [comment]
@bcemmett Oops.  Missed that.  
If I read the docs right,  [.Net 4.7](https://docs.microsoft.com/en-us/dotnet/api/system.net.servicepointmanager.securityprotocol?view=netframework-4.7.2#System_Net_ServicePointManager_SecurityProtocol) defaults to "SystemDefault" rather than "no default value".  Not sure if that relieves us from having to loop through a list of protocols, but it's something.

**Comment by cdru** [comment]
.Net 4.0 and higher will try to default to using TLS 1.0, 1.1, or 1.2 and not enabling SSL3. However, if ServicePointManager.DisableStrongCrypto is set to true, it'll force it down to TLS 1.0 or SSL3. DisableStrongCrypto is initialized based on if the value SchUseStrongCrypto data has been set to 0 or 1. 

If the value is missing, DisableStrongCrypto ends up defaulting to enabled (low security) for .Net 4.5.2 and below while 4.6.1 and higher end up defaulting to disabled (higher security).

You can set the key by running the two commands below for 32- and 64-bit installations. No recompiling of anything necessary. It's not without risks as you're still making the change globally, maybe even more so as you're now going beyond even the appdomain and affecting ALL .Net 4.x apps. But you probably know if you MUST keep SSL3 around for some ancient service.

Powershell commands (run with elevated prompt):
Set-ItemProperty -Path 'HKLM:\SOFTWARE\Wow6432Node\Microsoft\.NetFramework\v4.0.30319' -Name 'SchUseStrongCrypto' -Value '1' -Type DWord
Set-ItemProperty -Path 'HKLM:\SOFTWARE\Microsoft\.NetFramework\v4.0.30319' -Name 'SchUseStrongCrypto' -Value '1' -Type DWord

**Comment by victortorgrimson** [comment]
Thank you!  The ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12 line fixed the problem for me.  By the way, thank you for the great API, it has been very helpful!

**Comment by bcemmett** [comment]
Looked into this a little more. There are a couple of possible approaches:
- Set `ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;` (or alternatively OR it in).
- Change some registry keys (either @cdru's approach, or by [enabling TLS support at the OS level](https://docs.microsoft.com/en-us/windows-server/security/tls/tls-registry-settings#tls-12)).

In both cases, I think the change is too invasive to just do without checking. If other code in the application is hitting a different web service which requires older SSL settings, that would be broken by this change. Even an approach of changing the setting before each web request and then changing it back to the remembered setting afterwards would potentially break apps where other requests are being made simultaneously on different threads, in a nastily unpredictable way. The Google API .NET client folks [came to a similar conclusion](https://github.com/google/google-api-dotnet-client/issues/911).

An alternative is for this library to create a new appdomain and make its web requests from there to allow it to change that SecurityProtocol setting without impacting the rest of the app, but that comes with significant complexity / downsides which I don't think are justified.

Therefore the library will now (as of 3.2.181) detect this error and raise a special exception recommending that `ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;` is set elsewhere in the application, and linking to this issue for context.

**Comment by cdru** [comment]
I agree with your thinking @bcemmett. My suggestion was just a rehash of what Microsoft recommends for apps targeting versions prior to .Net 4.6 as a possible solution other than making people recompile existing code. It does have its drawbacks.

I'd suggest tweaking the error message slightly. How it's worded now still can cause unintended consequences within the app domain, especially if the developer doesn't read this issue discussion or understand it's effective for the entire app context. It also goes against points 2 and 3 of Microsoft's own [best practice guidelines](https://docs.microsoft.com/en-us/dotnet/framework/network-programming/tls) on the issue:
> Target .NET Framework 4.7 or later versions on your apps. Target .NET Framework 4.7.1 or later versions on your WCF apps.
> Do not specify the TLS version. Configure your code to let the OS decide on the TLS version.
> Perform a thorough code audit to verify you're not specifying a TLS or SSL version.

I think a better solution would be to just tell the developer of the error, and point them to this issue tracker and Microsoft's best practices for workarounds/solutions. So something like:
> SSL/TLS error. SurveyMonkey requires TLS 1.2, as of 13 June 2018. 
> See https://github.com/bcemmett/SurveyMonkeyApi-v3/issues/66 for discussion on issue. 
> See https://docs.microsoft.com/en-us/dotnet/framework/network-programming/tls for details with suggested workarounds and their drawbacks.

I'd also suggest adding it as a note or requirement on the project's readme.md.
